definitions/npm/async_v1.x.x/flow_>=v0.28.x/async_v1.x.x.js

```diff
@@ -0,0 +1,77 @@
+// @flow
+
+declare module 'async' {
+  declare type Collection<TItem> = Iterable<TItem> | { [key: string]: TItem };
+
+  declare type IterateeCallback<TError, TResult> = (err?: TError, result?: TResult) => void;
+  declare type Callback<TError, TResult> = (err?: TError, result?: TResult) => mixed;
+
+  // Collections
+  declare function each<TItem, TError>(
+    collection: Collection<TItem>,
+    iteratee: (item: TItem, callback: IterateeCallback<TError, void>) => mixed,
+    callback?: Callback<TError, void>
+  ): void;
+  declare function forEachOf<TItem, TError>(
+    collection: Collection<TItem>,
+    iteratee: (value: TItem, key: string | number, callback: IterateeCallback<TError, void>) => mixed,
+    callback?: Callback<TError, void>
+  ): void;
+  declare function map<TItem, TResult, TError>(
+    collection: Collection<TItem>,
+    iteratee: (value: TItem, callback: IterateeCallback<TError, TResult>) => mixed,
+    callback?: Callback<TError, Array<TResult>>
```

Comment:
In async.js the `error` in the final callback behaves a bit weird.

If you pass a **truthy** value to the iteratee callback, it will be passed to the final callback.
If you pass a **falsy** value to the iteratee callback, `null` will be passed to the final callback.

This is really hard to write a type annotation for. I tried declaring two separate methods, but then flow can't decide which to use.

I've created a simplified version of this: https://gist.github.com/thejameskyle/17e80cc7f2a6ff45d351c00185dd2db4